### PR TITLE
TEP-0061, Allow custom task to be embedded.

### LINKED ISSUE #3682: Custom Tasks: question with taskSpec
related #3463

The current custom task can run in a pipeline using `pipeline.taskRef`. Will there be any support for `pipeline.taskSpec` in the roadmap?

/cc @GregDritschler @fenglixa

### PR DESCRIPTION
**Pull request by ScrapCodes**
# Changes

## TEP-0061, Allow custom task to be embedded
    
    
 1. API changes, This PR adds new APIs i.e. adds a field `Spec *EmbeddedRunSpec` to `RunSpec`
 2. An embedded task will accept new field  `Spec` with type `runtime.RawExtension` in addition to `ApiVersion` and `Kind` fields of type string (as part of `runtime.TypeMeta`) :
 3. Validation changes, in addition to adding support for Run.RunSpec.Spec the validations will be changed to support "One of `Run.RunSpec.Spec` or `Run.RunSpec.Ref`" only and not both as part of a single API request to kubernetes.
 4. Unit tests + e2e test to cover the newly added fields and above mentioned validations.
 5. Documentation changes associated with TEP-0061.
References: 
1. https://github.com/tektoncd/community/pull/415
2. https://github.com/tektoncd/pipeline/issues/3682
3. https://github.com/tektoncd/community/blob/main/teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md

@Tomcli
@pritidesai 
@afrittoli 
@animeshsingh 

/cc @sbwsg @imjasonh

/kind feature

Closes #3682 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

```release-note

# It is now possible to embed the spec of a custom task in a Run resource, whether stand-alone or embedded in a Pipeline.

e.g.

          apiVersion: tekton.dev/v1alpha1
          kind: Run
          metadata:
            name: simpletasklooprun
          spec:
            params:
              - name: word
                value:
                  - jump
                  - land
              - name: suffix
                value: ing
            spec:
              apiVersion: custom.tekton.dev/v1alpha1
              kind: TaskLoop
              spec:
                # This is a new field, which comprises of custom-task spec.
                taskRef:
                  name: simpletask
                iterateParam: word
                timeout: 60s
                retries: 2


1. API changes, This PR adds new APIs i.e. adds a field `Spec *EmbeddedRunSpec` to `RunSpec`
2. An embedded task will accepts new field `Spec` with type `runtime.RawExtension` in addition to
    `ApiVersion` and `Kind` fields of type string (as part of `runtime.TypeMeta`) 
3. Validation changes, in addition to adding support for `Run.RunSpec.Spec` the validations will be changed
    to support "One of `Run.RunSpec.Spec` or `Run.RunSpec.Ref` " only and not both as part of a single
    API request to kubernetes.

action required: Developers of custom controllers (existing and new), who would like to support
 embedded spec for their custom task, need to implement the logic required to extract, validate
 and use the custom task spec from the new RunSpec.Spec field. Please review the documentation
 on upgrading, for more details and some examples.

```

### CODE REVIEW

**Comment by Tomcli** [comment]
@ScrapCodes looks like you haven't run `hack/update-codegen.sh` to regenerate the api code

**Comment by afrittoli** [comment]
> # Changes
> ##TEP-0061 (1/3), Allow custom task to be embedded.
> 
> ```
>     This is first in the three part series of PRs for implementing the TEP-0061.
> ```
> 
> 1. API changes, This PR adds new APIs i.e. adds a field `Spec *v1beta1.EmbeddedTask` to `RunSpec`
> 2. An embedded task will accept new field  `Spec` with type `runtime.RawExtension` in addition to `ApiVersion` and `Kind` fields of type string (as part of `runtime.TypeMeta`) :
> 3. Validation changes, in addition to adding support for Run.RunSpec.Spec the validations will be changed to support "One of `Run.RunSpec.Spec` or `Run.RunSpec.Ref`" only and not both as part of a single API request to kubernetes.
> 4. Unit tests to cover the newly added fields and above mentioned validations.
> 
> (2/3) Will consist of an e2e test.

To be honest, I would feel more comfortable having the E2E test part of this PR.
The PR is not too large, so it would be ok to have all in one go in fact.

> (3/3) Documentation changes associated with TEP-0061.

Again, I would prefer to have this in the same PR, but I'm ok with having it as a follow-up, as long as we merge both PR as part of the same release.

> 
> /kind feature
> 
> # Submitter Checklist
> As the author of this PR, please check off the items in this checklist:
> 
> * [ ]  [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
> * [x]  [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
> * [x]  Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
> * [x]  Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
>   functionality, content, code)
> * [x]  Release notes block below has been filled in or deleted (only if no user facing changes)
> 
> # Release Notes
> ```
>  1. API changes, This PR adds new APIs i.e. adds a field `Spec *v1beta1.EmbeddedTask` to `RunSpec`
>  2. An embedded task will accepts new field `Spec` with type `runtime.RawExtension` in addition to
>      `ApiVersion` and `Kind` fields of type string (as part of `runtime.TypeMeta`) 
>  3. Validation changes, in addition to adding support for `Run.RunSpec.Spec` the validations will be changed
>      to support "One of `Run.RunSpec.Spec` or `Run.RunSpec.Ref` " only and not both as part of a single
>     API request to kubernetes.
> ```

NIT: Could you add a more high level description on top or the release notes, something along the lines of 
"It is now possible to embed the spec of a custom task in a `Run` resource, whether stand-alone or embedded in a Pipeline. 
[include a yaml example]. Embedding the spec of a custom task depends on this being supported in each custom task controller that wants to support this feature."
 
> 
> ```
> action required: For developers of custom controllers (existing and new), now a `RunSpec`
>  can have a `Ref` or `Spec`, Scenario, where custom controller supports either or both `Ref`
>  and `Spec`. Please update the validation logic so that it responds with proper validation errors
>  or response to the user for missing a `Ref` and or `Spec`.
> ```

Could you keep this block ^^^ together with the release notes? Having two blocks might break the release notes automation.

I think it's a good idea to make a call for action, but we could say that "Developers of custom controllers (existing and new), who would like to support embedded spec for their custom task, need to implement the logic required to extract, validate and use the custom task spec from the new `RunSpec.Spec` field. "

As discussed during the API WG and noted by @imjasonh - having validation of the spec via a validating webhook would mean that all validating webhook of all custom tasks would be invoked for each `Run` - which might be too much. Perhaps we should recommend async validation during the reconcile cycle instead for now?

> 
> References:
> 
> 1. [tektoncd/community#415](https://github.com/tektoncd/community/pull/415)
> 2. #3682
> 3. https://github.com/tektoncd/community/blob/main/teps/0061-allow-custom-task-to-be-embedded-in-pipeline.md
> 
> @Tomcli
> @pritidesai
> @afrittoli
> @animeshsingh
> 
> /cc @sbwsg @imjasonh

**Comment by afrittoli** [comment]
Thanks a lot @ScrapCodes for this, I look forward to this feature!  I will review the code today.
We will have release v0.24 next week (it's planned for the 5th), do you think we include this in the milestone for the release?

**Comment by ScrapCodes** [comment]
> @ScrapCodes: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our release note process to remove it.

Looks like this robot does not interpret nested backticks. So, removed them.

**Comment by ScrapCodes** [comment]
Hi @afrittoli, thanks for taking a look !

Let us aim for 0.24 release, at the moment, only thing pending is review + e2e test. Do we have a code freeze process?

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (afrittoli) [inline pkg/reconciler/pipelinerun/pipelinerun.go:785]: I'm not sure I understand this. We set the `Run.Spec` in L763.
Why do we need to rebuild and `EmbeddedTask` from the `PipelineTask.TaskSpec` (which is an `EmbeddedTask` already)?

File: docs/runs.md
Comment (afrittoli) [inline docs/runs.md:101]: Isn't this already validated by the pipeline controller?

File: docs/runs.md
Comment (afrittoli) [inline docs/runs.md:104]: I'm not sure we need this?  It's up to the custom controller to decide what they need.

In Tekton we never create a resource for embedded specs - it does not add any value - or worst, it creates confusion, since it's not clear anymore which version of the resource is being executed, the one on the cluster or the one embedded.

So if we say something I think it should be that we discourage custom controllers from storing the embedded custom resource in `etcd`.

File: docs/runs.md
Comment (afrittoli) [inline docs/runs.md:118]: NIT: missing blank line on top of L118

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go
Comment (afrittoli) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:360]: removed the empty line?

**Review by afrittoli — COMMENTED** [review]
This is definitely in the right direction, thank you!
I left a couple of comments, mostly minor. 
I'm not sure about the `Run` creation logic in the pipeline run controller, but otherwise it looks good!

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (ScrapCodes) [reply pkg/reconciler/pipelinerun/pipelinerun.go:785]: My understanding is, we need to embed a `json.Marshal` of the `TaspSpec.Spec`. Please feel free to correct me.

File: docs/runs.md
Comment (ScrapCodes) [reply docs/runs.md]: You are right, it can be confusing and it is up to the custom controller, which way they want to implement.

**Comment by ScrapCodes** [comment]
Thank you @afrittoli , your feedback was very helpful, esp. the adding of `//+optional` to the API field and removing the redundant code from `pipelinerun.go` and your comments on the doc have been addressed hopefully. Please take a look again, when you can.

File: pkg/apis/pipeline/v1alpha1/run_validation_test.go
Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_validation_test.go:54]: done !

**Comment by Tomcli** [comment]
@afrittoli @pritidesai Thanks for the reviews. Do we need to also ask someone else to review? or we have to merge our TEP first?
https://github.com/tektoncd/community/pull/415

File: test/custom_task_test.go
Comment (sbwsg) [inline test/custom_task_test.go:72]: Suggest adding a `Spec` nested here and validating that it's rendered to `Raw` as expected

File: pkg/apis/pipeline/v1alpha1/run_types.go
Comment (sbwsg) [inline pkg/apis/pipeline/v1alpha1/run_types.go:46]: Ah, something I had not realised during TEP review is that this also means we're nesting `TaskSpec` fields (like pipeline resources etc,) and `PipelineTaskMetadata` fields (annotations, labels) into the `RunSpec`.

Is there a reason to re-use `EmbeddedTask` here?  Or could we introduce a new type that only includes the fields we want for `RunSpec`?  E.g.

```go
type EmbeddedRunSpec struct {
	// +optional
	runtime.TypeMeta `json:",inline,omitempty"`

	// Spec is a specification of a custom task
	// +optional
	Spec runtime.RawExtension `json:"spec,omitempty"`
}
```

Comment (sbwsg) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: Sorry, another question after reviewing this with the pipelines owners group - what is the purpose of `runtime.TypeMeta` in this location?  Would this potentially differ from the `run.apiVersion` and `run.kind`?

Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: >Sorry, another question after reviewing this with the pipelines owners group - what is the purpose of runtime.TypeMeta in this location? Would this potentially differ from the run.apiVersion and run.kind? 

It will not differ from `run.apiVersion` and `run.kind`, my guess is for brevity we prefer this.

>Is there a reason to re-use EmbeddedTask here? Or could we introduce a new type that only includes the fields we want for RunSpec? E.g.

Intention is, custom tasks will become consistent with regular tasks at some point. e.g. we need `PipelineTaskMetadata` to provide label to embedded task. 
If you still feel otherwise, let us know.

Comment (sbwsg) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: ## TypeMeta

Ah, ok.  Going back to the docs in this example helped clear things up for me wrt TypeMeta:

```
apiVersion: tekton.dev/v1alpha1
kind: Run
metadata:
  name: embedded-run
spec:
  spec:
    apiVersion: example.dev/v1alpha1
    kind: Example
    spec:
      field1: value1
      field2: value2
```

The run's API version and the spec's API version can differ - one describes the tekton resource "tekton.dev/v1alpha1.Run" and the other describes the Custom Task's own apiVersion and kind.  I find this distinction _endlessly_ confusing but that's just the way it is with Custom Tasks. Apologies for muddying things up here.  TypeMeta is fine as-is.

---

## EmbeddedTask

I do want to avoid using `EmbeddedTask` if we can for the following reasons:
- It brings baggage with it that we don't have a concrete use case for yet in Runs (e.g. pipeline resources).
- All of the fields from `TaskSpec` are already representable by virtue of using `runtime.RawExtension` I think?
- We can make `EmbeddedRunSpec` a subset of fields from `EmbeddedTask` and eventually converge them if that's a goal we have.  Although I would also say here that I wasn't aware of this being a Goal - is it mentioned in the TEP or a comment on it?  I probably missed it.

So to support metadata fields as well it sounds like the shape would become:

```go
type RunSpecMetadata struct {
	// +optional
	Labels map[string]string `json:"labels,omitempty"`

	// +optional
	Annotations map[string]string `json:"annotations,omitempty"`
}

type EmbeddedRunSpec struct {
	// +optional
	runtime.TypeMeta `json:",inline,omitempty"`

	// +optional
	Metadata RunSpecMetadata `json:"metadata,omitempty"`

	// Spec is a specification of a custom task
	// +optional
	Spec runtime.RawExtension `json:"spec,omitempty"`
}
```

The `createRun` func becomes a little more complicated because it has to copy values from the `EmbeddedTask` to the `EmbeddedRunSpec` but I much prefer we start with a constrained set of fields and expand it when needed.  WDYT?

Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: Alright, agree with you, do you think we would need to update the TEP with above suggestion for types?

Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: And get it approved?

Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: @sbwsg At the moment a `PipelineTask` has a nested `EmbeddedTask` and as a result, the custom task can be expressed under the field `taskSpec`. After this change, how will that change? 

Few options:
1. We introduce a new optional field `spec` to store the `EmbeddedRunSpec` inside a `PipelineTask`. Downside, two different field with similar sounding field names and purposes.
2. Leave current `EmbeddedTask` as is, and introduce a new type `EmbeddedRunSpec` for `Run`. Downside, same fields copied twice.
3. Any other suggestions?

Comment (sbwsg) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: Once this PR is merged the TEP can be moved to `"implemented"` state.  My suggestion would be to include these changes as part of that PR.

Also pinging @tektoncd/core-maintainers for any other opinions / feedback on the desired process here.

Comment (sbwsg) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: > Leave current EmbeddedTask as is, and introduce a new type EmbeddedRunSpec for Run. Downside, same fields copied twice.

This would definitely be my preference.

Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_types.go:46]: Yes, indeed.

Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_types.go]: @sbwsg One question, now that `Metadata` field is shifted from `EmbeddedTask` to `EmbeddedSpec` which is inlined into `EmbeddedTask`, is it a matter of concern?

File: internal/builder/v1beta1/pipeline_test.go
Comment (ScrapCodes) [inline internal/builder/v1beta1/pipeline_test.go:157]: Please note, those already using the PipelineTaskMetadata field in the EmbeddedTask, will have to make this change. 
//cc @sbwsg

**Comment by ScrapCodes** [comment]
Hello @sbwsg , hopefully this should address all of the comments. Can you please take a look again ! 😄

File: internal/builder/v1beta1/pipeline_test.go
Comment (sbwsg) [reply internal/builder/v1beta1/pipeline_test.go:157]: Ah I see, the problem becomes for users who are interacting with Tekton by importing its types - they now have to update their code to populate the embedded struct.  Is that right?

That's a really interesting point.  I think this is an argument in favor of keeping the two structs (`EmbeddedRunSpec` and `EmbeddedTask`) as totally separate copies in that case.

Comment (ScrapCodes) [reply internal/builder/v1beta1/pipeline_test.go:157]: ok, will update the code accordingly.

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (sbwsg) [inline pkg/reconciler/pipelinerun/pipelinerun.go:775]: 👍

**Review by sbwsg — APPROVED** [review]

**Comment by sbwsg** [comment]
Thanks a lot @ScrapCodes !

**Comment by animeshsingh** [comment]
Thanks @ScrapCodes! Critical feature for our use-cases

File: docs/pipelines.md
Comment (pritidesai) [inline docs/pipelines.md:1203]: Please add a reference to the TEP here so that its easier to relate to the proposal 🙏

File: docs/runs.md
Comment (pritidesai) [inline docs/runs.md:50]: the reference is invalid here, please replace it with #1-specifying-the-target-custom-task-with-ref

File: docs/runs.md
Comment (pritidesai) [inline docs/runs.md:52]: same here, please replace the reference with `#2-specifying-the-target-custom-task-by-embedding-its-spec`

File: docs/runs.md
Comment (pritidesai) [inline docs/runs.md:72]: same as above, the valid reference is `#1-specifying-the-target-custom-task-with-ref`

File: docs/runs.md
Comment (pritidesai) [inline docs/runs.md:75]: ditto 🙃 valid reference here is ` #2-specifying-the-target-custom-task-by-embedding-its-spec`

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/pipelinerun.go:782]: 🎉

File: pkg/apis/pipeline/v1alpha1/run_types.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1alpha1/run_types.go:42]: We consider this as a required field so do we need to add `omitempty` clause here?

File: pkg/apis/pipeline/v1alpha1/run_validation_test.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1alpha1/run_validation_test.go:47]: => empty ref and spec

File: pkg/apis/pipeline/v1alpha1/run_validation.go
Comment (pritidesai) [inline pkg/apis/pipeline/v1alpha1/run_validation.go:44]: need an additional check here, if both are `nil`

File: pkg/apis/pipeline/v1alpha1/run_validation_test.go
Comment (pritidesai) [reply pkg/apis/pipeline/v1alpha1/run_validation_test.go:47]: the test name does not match with ref and spec both being `nil`, please add one more check in the validation routine and update this test.

File: pkg/controller/filter.go
Comment (pritidesai) [inline pkg/controller/filter.go:56]: Nothing wrong with your changes here, just an observation:

`r.Spec.Ref.Kind` is of type `TaskKind` and that's why needs the conversion v/s `r.Spec.Spec.Kind` is of type `string`. `TaskKind` is also originally used to signify whether a task is `task` or `clustertask` but with custom task, it can be set to anything.

File: pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go:366]: Please separate this entire logic into a separate function to check whether a task is a custom task and assign `rprt.CustomTask` appropriate value here based on that new function. `ResolvePipelineRunTask` will be simplified instead of making everything part of it.

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (pritidesai) [inline pkg/reconciler/pipelinerun/pipelinerun.go:771]: what happens when `spec` is empty here? do we need a check here `rprt.PipelineTask.TaskSpec.Spec`?

**Comment by pritidesai** [comment]
the changes looks great @ScrapCodes. I am done reviewing the PR and have requested couple of changes. 🙏 

One more thing I am not sure about is all the third party changes, without changing the `mod` file, how come we are seeing the dependency change 🤔

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (ScrapCodes) [reply pkg/reconciler/pipelinerun/pipelinerun.go:771]: there is no error for empty spec, in fact a non empty byte sequence is generated. Do you think we should be more defensive here?

File: pkg/apis/pipeline/v1alpha1/run_validation_test.go
Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_validation_test.go:47]: The reason is, both nil and `Empty spec` are same in terms of what happens underneath. 
Actual check says, 
`if equality.Semantic.DeepEqual(rs, &amp;RunSpec{}) {		return apis.ErrMissingField("spec")` i.e. empty spec and so is the name. Please see the reply to your comment above as well, and let me know, if this is still concerning.

**Comment by ScrapCodes** [comment]
> the changes looks great @ScrapCodes. I am done reviewing the PR and have requested couple of changes. 🙏
> 
> One more thing I am not sure about is all the third party changes, without changing the `mod` file, how come we are seeing the dependency change 🤔

I am not sure either, they are generated as part of running, `hack/update-codegen.sh`.

**Comment by ScrapCodes** [comment]
Looks like it is a problem with my local environment. Even after removing those `third_party` additions, build passes i.e. build passes regardless of those `third_party` additions are present or not. Is this concerning from the point of view, how build is setup.

**Comment by ScrapCodes** [comment]
Thank you @pritidesai for taking a look. Please take a look again.

Thank you @sbwsg !

Thank you @animeshsingh and @Tomcli for all the support 😄

File: pkg/reconciler/pipelinerun/pipelinerun.go
Comment (pritidesai) [reply pkg/reconciler/pipelinerun/pipelinerun.go:771]: I am trying to think if there is any use case where the `spec` could be empty, could this ever be used in any way? 

Since this is not producing any error, its safe to go with the way its implemented now and revisit if needed in future.

**Comment by Tomcli** [comment]
thanks @pritidesai for your review

File: pkg/apis/pipeline/v1alpha1/run_validation.go
Comment (ScrapCodes) [reply pkg/apis/pipeline/v1alpha1/run_validation.go:44]: Thanks, now I have understood ! Will open a PR with the fix soon.

**Comment by ScrapCodes** [comment]
@pritidesai I have opened a follow-up PR with the validation: https://github.com/tektoncd/pipeline/pull/3977

**Pull request merged**
